xc/device.ml

```diff
@@ -1143,18 +1143,67 @@ module PCI = struct
     (* Sort into the order the devices were plugged *)
     List.sort (fun a b -> compare (fst a) (fst b)) pairs
 
-  let add ~xs pcidevs domid =
+  let _pci_add ~xc ~xs pcidev domid slot =
+    let sysfs_pci_dev = "/sys/bus/pci/devices/" in
+    let bus = 0 in
+    let func = 0 in
+    let pci_bdf = Printf.sprintf "%04x:%02x:%02x.%01x" domid bus slot func in
+    let page_size = 4096 in
+    if (Qemu.is_running ~xs domid) then
+      begin
+        let _qmp_result = qmp_send_cmd domid
+            (Qmp.Device_add {driver="xen-pci-passthrough";
+                             device=Qmp.Device.PCI({id=pcidev; dev=0; fn=0; hostaddr=pcidev; permissive=false})}) in
+        Unixext.with_input_channel (sysfs_pci_dev ^ pci_bdf ^ "/resource") (fun in_channel ->
+            for i = 0 to 7 do
+              let start = ref 0 in
+              let end_ = ref 0 in
+              let flags = ref 0 in
+              let size = ref 0 in
+              Scanf.bscanf (Scanf.Scanning.from_channel in_channel) "0x%llx 0x%llx 0x%llx\n" (fun scan_start scan_end scan_flags ->
+                  start := scan_start;
```

Comment:
Instead of using all these refs and `<> 0` checks, why not do the xenctrl call inside this lambda function?

Reply:
I can remove the refs, and move the uses of them into the lambda function, but I think I still need the non-zero check